src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import akka.http.scaladsl.model.StatusCodes
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.plus.PlusScopes
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.gson.internal.LinkedTreeMap
+import com.typesafe.config.Config
+import io.swagger.client.ApiClient
+import io.swagger.client.api.ResourcesApi
+import io.swagger.client.api.GoogleApi
+import java.io.{ByteArrayInputStream, File}
+import java.util.concurrent.TimeUnit
+import com.google.gson.Gson
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model._
+import org.broadinstitute.dsde.workbench.leonardo.model.Actions._
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions._
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions._
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import net.ceedubs.ficus.Ficus._
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+case class UserEmailAndProject(userEmail: String, googleProject: String)
```

Comment:
The parameter types here could be `WorkbenchEmail` and `GoogleProject`.

Also should we put this in `LeonardoModel`?

Reply:
they're strings in the rest of the interface in order to minimize dependencies. doesn't using the value classes just introduce unnecessary applying to and fro?

(that thread that went by in takeover_plot about how to do value classes more properly is interesting in this context)

Reply:
We use value classes in the rest of Leo for better type checking. The value classes in Leo are true Scala value classes (e.g. they extend `AnyVal`) in that they have 0 runtime overhead; the ones in wb-libs are just case class wrappers which incur runtime overhead, but it's probably minimal (housekeeping: convert wb-libs value classes to extend `AnyVal` :)).

The dependency on wb-libs is there regardless; Leo requires it in order to work. I guess the question is whether we want to hide all wb-libs _compile-time_ dependencies from the provider classes, to make development easier. That sounds reasonable to me, but we should:
- add a comment explaining that to the interfaces
- clean up all method signatures to not use value classes (looking at current code, I see a mix of value classes and `String`s)

Reply:
agree that we should make a decision one way or the other. i've been wavering. probably a good idea to talk to end-user implementers (i.e. verily devs) and see which they'd prefer

Reply:
What should we do in the interim? Use the model classes or the strings?